File: backend/src/services/zipService.ts
import { readFile } from "node:fs/promises";
import { zipSync, type Zippable } from "fflate";

export interface ZipEntry {
  /** Path/name the file should have inside the zip archive. */
  zipPath: string;
  /** Absolute filesystem path to read the file's bytes from. */
  absoluteFilePath: string;
}

/**
 * Reads every entry's bytes from disk and zips them into a single in-memory
 * buffer, synchronously (fflate's zipSync — the generated solution files are
 * small text files, so there's no need for the streaming/async zip API the
 * frontend's zipAndDownload.ts doesn't need either).
 */
export async function buildZip(entries: ZipEntry[]): Promise<Uint8Array> {
  const files: Zippable = {};
  for (const entry of entries) {
    files[entry.zipPath] = await readFile(entry.absoluteFilePath);
  }
  return zipSync(files);
}

File: backend-py/app/services/zip_service.py
"""Port of `backend/src/services/zipService.ts`, using Python's stdlib
`zipfile` instead of `fflate` — same "read every entry's bytes from disk,
zip them into one in-memory buffer" behavior."""

from __future__ import annotations

import zipfile
from dataclasses import dataclass
from io import BytesIO


@dataclass
class ZipEntry:
    """`zipPath`: path/name the file should have inside the zip archive.
    `absoluteFilePath`: absolute filesystem path to read the file's bytes
    from."""

    zipPath: str
    absoluteFilePath: str


def build_zip(entries: list[ZipEntry]) -> bytes:
    buffer = BytesIO()
    with zipfile.ZipFile(buffer, mode="w", compression=zipfile.ZIP_DEFLATED) as zf:
        for entry in entries:
            with open(entry.absoluteFilePath, "rb") as f:
                zf.writestr(entry.zipPath, f.read())
    return buffer.getvalue()
